### PR TITLE
Make HTTP/1.0 responses without Connection: keep-alive mean close.

### PR DESCRIPTION
**Pull request by jplevyak**
Signed-off-by: John Plevyak <jplevyak@gmail.com>

For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/master/PULL_REQUESTS.md)

Description: Currently HTTP/1.0 requests without a Connection: close cause the connection to remain open.  This does not conform to the HTTP/1.0 specification.  Connections without Connection: keep-alive should be expected to close.  Failing to do so can result in 503 errors if a new request is sent on that connection before the close is detected.
Risk Level: Medium
Testing: 24 hours with 5 applications
Docs Changes:
Release Notes: HTTP/1.0 responses without Connection: keep-alive now means close.
[Optional Fixes #Issue] https://github.com/envoyproxy/envoy/issues/7123
[Optional Deprecated:]

### CODE REVIEW

File: source/common/http/http1/conn_pool.cc [hunk 1]
```diff
@@ -203,7 +203,9 @@ void ConnPoolImpl::onResponseComplete(ActiveClient& client) {
   if (!client.stream_wrapper_->encode_complete_) {
     ENVOY_CONN_LOG(debug, "response before request complete", *client.codec_client_);
     onDownstreamReset(client);
-  } else if (client.stream_wrapper_->saw_close_header_ || client.codec_client_->remoteClosed()) {
+  } else if (client.stream_wrapper_->saw_close_header_ || client.codec_client_->remoteClosed() ||
+             (client.codec_client_->protocol() == Protocol::Http10 &&
+              !client.stream_wrapper_->saw_keep_alive_header_)) {
     ENVOY_CONN_LOG(debug, "saw upstream connection: close", *client.codec_client_);
```
Comment (alyssawilk) [inline source/common/http/http1/conn_pool.cc]: Maybe rephrase this log?

File: source/common/http/http1/conn_pool.cc [hunk 2]
```diff
@@ -273,11 +275,15 @@ ConnPoolImpl::StreamWrapper::~StreamWrapper() {
...
-    saw_close_header_ = true;
-    parent_.parent_.host_->cluster().stats().upstream_cx_close_notify_.inc();
+  if (headers->Connection()) {
+    if (absl::EqualsIgnoreCase(headers->Connection()->value().getStringView(),
+                               Headers::get().ConnectionValues.Close)) {
+      saw_close_header_ = true;
+      parent_.parent_.host_->cluster().stats().upstream_cx_close_notify_.inc();
+    } else if (absl::EqualsIgnoreCase(headers->Connection()->value().getStringView(),
+                                      Headers::get().ConnectionValues.KeepAlive)) {
+      saw_keep_alive_header_ = true;
```
Comment (alyssawilk) [inline source/common/http/http1/conn_pool.cc]: Optional but in the http connection manager we treat lack of keepalive for HTTP/1.0 as equivalent to explicit connection: close for HTTP/1.1 (i.e. use the same boolean) to make sure anywhere we special case behavior for one we special case it for the other.

I'm fine with you using separate booleans but I do think you should increment the upstream_cx_close_notify_ stat for this case (as it already is reused for GOAWAY) at which point you have to do protocol checks in this function at which point it might be nice avoiding the complexity in onResponseComplete

File: test/common/http/http1/conn_pool_test.cc [hunk 3]
```diff
@@ -537,7 +551,7 @@ TEST_F(Http1ConnPoolImplTest, ConnectionCloseWithoutHeader) {
   conn_pool_.test_clients_[0].connection_->raiseEvent(Network::ConnectionEvent::Connected);
 
   callbacks2.outer_encoder_->encodeHeaders(TestHeaderMapImpl{}, true);
-  response_headers.reset(new TestHeaderMapImpl{{":status", "200"}});
+  response_headers.reset(new TestHeaderMapImpl{{":status", "200"}, {"connection", "keep-alive"}});
```
Comment (alyssawilk) [inline test/common/http/http1/conn_pool_test.cc]: Do we need to change these tests?  If so, I'm a bit concerned the behavioral change might be larger scope than we intend.

File: source/common/http/http1/conn_pool.cc (hunk 1 above)
Comment (jplevyak) [reply source/common/http/http1/conn_pool.cc]: Even before it was conflating RemoteClose.  Changed to:

 ENVOY_CONN_LOG(debug, "saw upstream close connection", *client.codec_client_);

File: source/common/http/http1/conn_pool.cc (hunk 2 above)
Comment (jplevyak) [reply source/common/http/http1/conn_pool.cc]: Concretely are you suggesting combining the booleans into saw_upstream_close_ or some similar which is either Connection: close or for HTTP/1.0 the lack of the keep-alive header ?

Conversely I could move the upstream_cx_close_notify_ into onResponseComplete.

File: test/common/http/http1/conn_pool_test.cc (hunk 3 above)
Comment (jplevyak) [reply test/common/http/http1/conn_pool_test.cc]: Unfortunately the tests do not specify HTTP/1.1 and assume that the lack of a Connection: close means keep-alive.  We could change them to be HTTP/1.1 or add the keep-alive. The protocol isn't explicitly specified so it might just be an oversight and in any case the protocol will be explicit in non-test operation.   Suggestion?

Comment (alyssawilk) [reply test/common/http/http1/conn_pool_test.cc]: Could default mock out the codec to return something other than 0 for protocol?  If that's not trivial I'm OK with this as-is - poked through the code and sorted out why it's the case.  As long as we didn't need to tweak integration tests I'm still happy it's low impact.

File: source/common/http/http1/conn_pool.cc (hunk 2 above)
Comment (jplevyak) [reply source/common/http/http1/conn_pool.cc]: The protocol isn't available in decodeHeaders() so I moved the stat increment into onResponseComplete which is consistent with the log message and the increment of upstream_cx_max_requests_.   PTAL

File: test/common/http/http1/conn_pool_test.cc (hunk 3 above)
Comment (jplevyak) [reply test/common/http/http1/conn_pool_test.cc]: The tests run at a higher layer (decodeHeaders) rather than running the parser_, so I would have to just override the protocol() which would mean that the behavior would not match any underlying data and the CodecClient::protocol() method is not virtual currently.  Also that would mean changing the test from HTTP/1.0 to HTTP/1.1.   I agree that it would be better to have these tests for both HTTP/1.0 and HTTP/1.1 and that the author might have thought that they were for 1.1 but they have been 1.0 so perhaps these should stay 1.0 and we can add some more 1.1 tests. 

Created https://github.com/envoyproxy/envoy/issues/7158

File: source/common/http/http1/conn_pool.cc (hunk 2 above)
Comment (mattklein123) [reply source/common/http/http1/conn_pool.cc]: +1 to Alyssa's suggestion of having a single boolean track whether we want to close the connection after processing headers.

File: source/common/http/http1/conn_pool.cc [hunk 4]
```diff
@@ -273,17 +276,19 @@ ConnPoolImpl::StreamWrapper::~StreamWrapper() {
 void ConnPoolImpl::StreamWrapper::onEncodeComplete() { encode_complete_ = true; }
 
 void ConnPoolImpl::StreamWrapper::decodeHeaders(HeaderMapPtr&& headers, bool end_stream) {
-  if (headers->Connection() &&
-      absl::EqualsIgnoreCase(headers->Connection()->value().getStringView(),
-                             Headers::get().ConnectionValues.Close)) {
-    saw_close_header_ = true;
-    parent_.parent_.host_->cluster().stats().upstream_cx_close_notify_.inc();
+  if (headers->Connection()) {
+    if (absl::EqualsIgnoreCase(headers->Connection()->value().getStringView(),
```
Comment (mattklein123) [inline source/common/http/http1/conn_pool.cc]: perf nit: can you check protocol in each of these if cases so we short circuit the connection check depending on the protocol?

File: test/common/http/http1/conn_pool_test.cc (hunk 3 above)
Comment (mattklein123) [reply test/common/http/http1/conn_pool_test.cc]: +1 please don't change the existing tests, or clearly mark them as HTTP/1.0 and we can add HTTP/1.1 tests. Let's figure out how to write new tests that cover the HTTP/1.0 case, so let's resolve the issue you opened as part of this PR.

File: source/common/http/http1/conn_pool.cc (hunk 2 above)
Comment (jplevyak) [reply source/common/http/http1/conn_pool.cc]: The protocol isn't available in decodeHeaders().  Without it I can't have a single variable.  Suggestions?

File: test/common/http/http1/conn_pool_test.cc (hunk 3 above)
Comment (jplevyak) [reply test/common/http/http1/conn_pool_test.cc]: I renamed the tests which depend on HTTP/1.0 behavior to include Http10 as a prefix.

File: source/common/http/http1/conn_pool.cc (hunk 2 above)
Comment (mattklein123) [reply source/common/http/http1/conn_pool.cc]: You can get the `protocol()` from the codec, which is updated prior to header callbacks.

/wait

Comment (jplevyak) [reply source/common/http/http1/conn_pool.cc]: Done.

File: source/common/http/http1/conn_pool.cc (hunk 4 above)
Comment (jplevyak) [reply source/common/http/http1/conn_pool.cc]: The logic has changed with the protocol checks.  PTAL

File: source/common/http/http1/conn_pool.cc [hunk 5]
```diff
@@ -273,17 +273,20 @@ ConnPoolImpl::StreamWrapper::~StreamWrapper() {
...
+        !absl::EqualsIgnoreCase(headers->Connection()->value().getStringView(),
+                                Headers::get().ConnectionValues.KeepAlive)))) {
     parent_.parent_.host_->cluster().stats().upstream_cx_close_notify_.inc();
-  }
-  if (!saw_close_header_ && headers->ProxyConnection() &&
-      absl::EqualsIgnoreCase(headers->ProxyConnection()->value().getStringView(),
-                             Headers::get().ConnectionValues.Close)) {
-    saw_close_header_ = true;
+    close_connection_ = true;
+  } else if (headers->ProxyConnection() &&
```
Comment (mattklein123) [inline source/common/http/http1/conn_pool.cc]: Can you fold this into the previous if statement? I think it could be part of the same or statement. Feel free to split the logic out into const bools if it helps readability of the final statement. Mainly I would like the stat incremented in only one place (I realize you didn't do this in the first place but might as well clean up).

File: test/common/http/http1/conn_pool_test.cc [hunk 6]
```diff
@@ -269,13 +281,13 @@ TEST_F(Http1ConnPoolImplTest, VerifyBufferLimits) {
  * Tests a request that generates a new connection, completes, and then a second request that uses
  * the same connection.
  */
-TEST_F(Http1ConnPoolImplTest, MultipleRequestAndResponse) {
+TEST_F(Http1ConnPoolImplTest, Http10MultipleRequestAndResponse) {
```
Comment (mattklein123) [inline test/common/http/http1/conn_pool_test.cc]: I think it would be better if you defaulted protocol in the codec to 1.1, and left all the existing tests alone, and then added some new 1.0 tests.

File: test/common/http/http1/conn_pool_test.cc (hunk 3 above)
Comment (alyssawilk) [reply test/common/http/http1/conn_pool_test.cc]: I think Matt's ask was that we either have 1.0 tests and 1.1 tests, or simply 1.1 tests, and while you did the rename we don't have the HTTP/1.1 case covered

So you can either copy/paste/edit the 1.0 tests to 1.1 tests, or go for minimal changes, say making expectClientCreate(Protocol protocol =  Protocol::Http11) ) (in your new test have it be HTTP/1.0) and adding
ON_CALL(test_client.codec_, protocol()).WillByDefault(Return(protocol)) to expectClientCreate

File: test/common/http/http1/conn_pool_test.cc (hunk 6 above)
Comment (jplevyak) [reply test/common/http/http1/conn_pool_test.cc]: Done.

File: test/common/http/http1/conn_pool_test.cc (hunk 3 above)
Comment (jplevyak) [reply test/common/http/http1/conn_pool_test.cc]: Done.

File: source/common/http/http1/conn_pool.cc (hunk 5 above)
Comment (jplevyak) [reply source/common/http/http1/conn_pool.cc]: Done

Comment (mattklein123) [reply source/common/http/http1/conn_pool.cc]: I don't see this one done? Did you mess a push?

/wait

Comment (jplevyak) [reply source/common/http/http1/conn_pool.cc]: Thanx, pushed.
